Wangscape/utils.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#include <type_traits>
+#include <stdexcept>
+
+namespace
+{
+
+template<typename I>
+I _pow(I base, I exp)
+{
+    I result = 1;
+    for (int i = 0; i < std::numeric_limits<I>::digits; i++)
+    {
+        if (exp & 1)
+            result *= base;
+        else if (exp == 0)
+            break;
+        exp >>= 1;
+        base *= base;
+    }
+    return result;
+}
+
+}
+
+template <typename I>
+typename std::enable_if<std::is_unsigned<I>::value, I>::type pow(I base, I exp)
+{
+    if (exp == 0)
+        return 1;
+    else if (base == 0)
+        return 0;
+    else if (base == 1)
+        return 1;
```

Comment:
For `long long` that will be **only** 63 iterations, I don't think it's a big problem, cause it shouldn't happen too often. Anyway, `base == 0` and `base == 1` are not unsigned types specific, so, if I'm right,  moving these checks to `_pow` won't break anything and will make it shorter and consistent.